# Week 02 Coding: Numpy, Pandas, and MatPlotLib (Lab)

## BEENOS Inc. Machine Learning Study Group

In the second week, you'll practice using `numpy`, `pandas` and `matplotlib`.

In [1]:
# add 2 to every element of a `list`

mylist = [1, 2, 3, 4, 5]

mylist + 2

TypeError: can only concatenate list (not "int") to list

But such an operation is trivial for `numpy`:

In [2]:
# add 2 to every element of an `ndarray`

import numpy as np

myarray = np.array([1,2,3,4,5])

myarray + 2

array([3, 4, 5, 6, 7])

We can also use `numpy` to create matrices from lists of lists:

In [3]:
# create a 3x3 matrix from a list of list

morelists = [[1,2,3], [4,5,6], [7,8,9]]

mymatrix = np.array(morelists)

mymatrix

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

### `numpy` Practice

For the first part of this lab, you'll be completing short exercises to get you used to working with `numpy` arrays.

The exercises are taken from [Machine Learning Plus](https://www.machinelearningplus.com/python/101-numpy-exercises-python/) and increase in difficulty from Level 1 to Level 4.

Feel free to skip around to exercises that challenge you.

Good luck!

**Q.0** Import numpy as `np` and print the version.

*Sample Output*:

> `1.15.4`

In [ ]:
# INSERT CODE HERE

**Q.1** Use `np.array` to create a 1D array of numbers from 0 to 9.

*Output:*

> array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [ ]:
# INSERT CODE HERE

**Q.2** Extract all odd numbers from `arr`.

*Output:*

> array([1, 3, 5, 7, 9])

In [ ]:
arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

# INSERT CODE HERE

**Q.3** Use `reshape` to convert a 1D array to a 2D array with two rows.

*Output:*

>     array([[0, 1, 2, 3, 4],
>            [5, 6, 7, 8, 9]])

In [ ]:
arr = np.arange(10)

# INSERT CODE HERE

**Q.4** Use `np.full` to create an nD boolean array. Choose your size and `True` or `False`.

*Sample Output:*

>     array([[ True,  True],
>            [ True,  True],
>            [ True,  True]])

In [ ]:
# INSERT CODE HERE

**Q.5** Use `np.where` to replace all odd numbers in `arr` with `-1`, but **DO NOT** change `arr`.

*Sample Output*:

> array([ 0, -1,  2, -1,  4, -1,  6, -1,  8, -1]) `# newarray`

> array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]) `# arr`

In [ ]:
arr = np.arange(10)

# INSERT CODE HERE

print(arr, newarray)

**Q.6** Create the following pattern. **Use numpy functions.**

*Output:*

> array([1, 1, 1, 2, 2, 2, 3, 3, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3])

In [ ]:
a = np.array([1,2,3])

# INSERT CODE HERE

**Q.7** Get all common items in both `a` and `b`.

*Output:*

> array([2,4])

In [ ]:
a = np.array([1,2,3,2,3,4,3,4,5,6])
b = np.array([7,2,10,2,7,4,9,4,9,8])

# INSERT CODE HERE

**Q.8** Remove from `a` all items in `b`.

*Output:*

> array([0, 1, 2, 4])

In [ ]:
a = np.array([0,1,2,3,4,5])
b = np.array([3,5,6,7,8,9])

# INSERT CODE HERE

**Q.9** Create a 2D array of size 5x3 containing random numbers between 5 and 10.

In [ ]:
# INSERT CODE HERE

**Q.10** Use `np.genfromtxt` to import the Iris flowers dataset from the given url.

*Hint: If your last column shows `nan` values, trying specifying an `object` data type.*

*Output:*

> `# Print the first row`

> array([[b'5.1', b'3.5', b'1.4', b'0.2', b'Iris-setosa']], dtype=object)

In [ ]:
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

# INSERT CODE HERE

## Part 1: Pandas

[CHANGE TO PANDAS]
NumPy is short for Numerical Python, and it is the most important package one needs when working with data in Python.

Other popular packages, like `pandas`, depend on the versatile functionality provided by numpy.

The core benefit to using `numpy` is access to its `ndarray` objects.

`ndarray` is short for **n-dimensional arrays**, which allow us to perform vectorized mathematical operations and data manipulation on collections of data.

### `pandas` Practice

For the next part of this lab, you'll be completing short exercises to get you used to working with `numpy` arrays.

The exercises are taken from [Machine Learning Plus](https://www.machinelearningplus.com/python/101-pandas-exercises-python/) and increase in difficulty from Level 1 to Level 3.

Feel free to skip around to exercises that challenge you.

Good luck!

**Q.0** Pandas relies on `numpy`. You should have already imported `numpy`. If not, import it and also import pandas (by convention as `pd`) and check your `pandas` version.

*Sample Output*:

> `0.23.4`

In [ ]:
# INSERT CODE HERE

In [5]:
import pandas as pd
import numpy as np

pd.__version__

'0.23.4'

**Q.1** Like `numpy`, `pandas` has popular structures that it is well known for. The first one is called a **Series**, which is a one-dimensional array. In laymans terms, a Series is simply a *column* of data, like one might find in an Excel spreadsheet.

A Series can hold any data type and can be coerced from multiple data structures. To test this, convert each of the following data structures (a list, an array and a dictionary) into a separate **Series**.

*Sample Output*:

> `# Print the head of the mydict series`

>     a     0
>     b     1
>     c     2
>     d     3
>     e     4
>     dtype: int64

In [29]:
mylist = list('abcdefghijklmnopqrstuvwyxz')
myarray = np.arange(1, 27)
mydict = dict(zip(mylist,myarray))

mylist_to_series = pd.Series(mylist) # INSERT CODE HERE
myarray_to_series = pd.Series(myarray) # INSERT CODE HERE
mydict_to_series = pd.Series(mydict) # INSERT CODE HERE

mydict_to_series.head()

a    1
b    2
c    3
d    4
e    5
dtype: int64

**Q.2** Series can also be coerced into the second fundamental `pandas` data structure, the **DataFrame**. It is a two-dimensional array made up of Series. Again, in laymans terms, a DataFrame is like one spreadsheet in Excel: it contains multiple columns.

Each column of a DataFrame cna hold any data type, and can be created from Series or other data structures. Let's try this by turning your `mydict` series into a `pandas` DataFrame. Use `.reset_index()` to add *both* columns to the table.

*Sample Output*:

> `# Print the head of the dataframe`

>           index  0
>     0     a      0
>     1     b      1
>     2     c      2
>     3     d      3
>     4     e      4

In [30]:
myseries = pd.Series(mydict)

mydataframe = myseries.to_frame().reset_index() # INSERT CODE HERE

mydataframe.head()

index  0
0     a  1
1     b  2
2     c  3
3     d  4
4     e  5

**Q.3** Our new dataframe looks a bit clunky. One problem is the column names `"index"` and `"0"` don't tell us much about the data in the table.

It might be easier to read if we change those names to `"letter"` and `"position"`, or even `"number"`. You can use `.rename()`.

*Hint: Does your head show `'index'` and `'0'` still? Try using `inplace`!*

*Sample Output*:

> `# Print the head of the dataframe`

>           letter  number
>     0     a       1
>     1     b       2
>     2     c       3
>     3     d       4
>     4     e       5

In [36]:
mydataframe.rename(columns={"index":"letter", 0:"number"}, inplace=True) #INSERT CODE HERE

mydataframe.head()

letter  number
0      a       1
1      b       2
2      c       3
3      d       4
4      e       5

**Q.4** Alright, let's practice with a bit of real data. We'll be looking at the [Cars93](https://raw.githubusercontent.com/selva86/datasets/master/Cars93_miss.csv) dataset for a few problems.

The dataset has already been loaded for you. Figure out how many rows and columns it contains.

*Sample Output*:

> (93, 27)

In [79]:
cars_df = pd.read_csv("https://raw.githubusercontent.com/selva86/datasets/master/Cars93_miss.csv")

cars_df.shape # INSERT CODE HERE

(93, 27)

**Q.5** Let's get a bit more info on our dataset. First, we want to figure out the *type* of the data that exists within each column. (What are we working with? Strings? Integers? Floats? Datetime objects?)

This is extremely important to know for manipulating our dataset later, because **all the data in a pandas *Series* or a DataFrame *column* must be the same**.

What types are in the Cars93 dataset?

*Sample Output (first 3 data types)*:

>     Manufacturer    object
>     Model           object
>     Type            object
>     dtype: object

In [80]:
cars_df.dtypes[:3] # INSERT CODE HERE

Manufacturer    object
Model           object
Type            object
dtype: object

**Q.6** As you work with real data, you'll soon find out that the more you *know* about the dataset you're working with, the easier it is to build models and make predictions.

One of the things you'll want to know is **summary statistics**: things like mean, standard deviation, maximum and minimum values, etc. Knowing these stats will help you determine whether your data needs to undergo transformation (like scaling and normalization) before modeling.

Show the summary stats for this dataset now. What is each row telling you?

*Sample Output*:

> `# First few columns of the first row`

>               Min.Price       Price           Max.Price
>     count     86.000000       91.000000       88.000000       ...
>     mean      17.118605       19.616484       21.459091       ...
>     ...

In [81]:
cars_df.describe() # INSERT CODE HERE

Min.Price      Price  Max.Price   MPG.city  MPG.highway  EngineSize  \
count  86.000000  91.000000  88.000000  84.000000    91.000000   91.000000   
mean   17.118605  19.616484  21.459091  22.404762    29.065934    2.658242   
std     8.828290   9.724280  10.696563   5.841520     5.370293    1.045845   
min     6.700000   7.400000   7.900000  15.000000    20.000000    1.000000   
25%    10.825000  12.350000  14.575000  18.000000    26.000000    1.800000   
50%    14.600000  17.700000  19.150000  21.000000    28.000000    2.300000   
75%    20.250000  23.500000  24.825000  25.000000    31.000000    3.250000   
max    45.400000  61.900000  80.000000  46.000000    50.000000    5.700000   

       Horsepower          RPM  Rev.per.mile  Fuel.tank.capacity  Passengers  \
count   86.000000    90.000000     87.000000           85.000000   91.000000   
mean   144.000000  5276.666667   2355.000000           16.683529    5.076923   
std     53.455204   605.554811    486.916616            3.375748    1.045953   
min     55.000000  3800.000000   1320.000000            9.200000    2.000000   
25%    100.750000  4800.000000   2017.500000           14.500000    4.000000   
50%    140.000000  5200.000000   2360.000000           16.500000    5.000000   
75%    170.000000  5787.500000   2565.000000           19.000000    6.000000   
max    300.000000  6500.000000   3755.000000           27.000000    8.000000   

           Length   Wheelbase      Width  Turn.circle  Rear.seat.room  \
count   89.000000   92.000000  87.000000    88.000000       89.000000   
mean   182.865169  103.956522  69.448276    38.954545       27.853933   
std     14.792651    6.856317   3.778023     3.304157        3.018129   
min    141.000000   90.000000  60.000000    32.000000       19.000000   
25%    174.000000   98.000000  67.000000    36.000000       26.000000   
50%    181.000000  103.000000  69.000000    39.000000       27.500000   
75%    192.000000  110.000000  72.000000    42.000000       30.000000   
max    219.000000  119.000000  78.000000    45.000000       36.000000   

       Luggage.room       Weight  
count     74.000000    86.000000  
mean      13.986486  3104.593023  
std        3.120824   600.129993  
min        6.000000  1695.000000  
25%       12.000000  2647.500000  
50%       14.000000  3085.000000  
75%       16.000000  3567.500000  
max       22.000000  4105.000000

**Q.7** Let's use our summary statistics to learn more about our data.

Find the `manufacturer`, `model` and `type` of car that has the highest `Price`. Use `.loc[]` to pull out this information.

*Sample Output*:

>               Manufacturer    Model     Type       Price
>     58        Mercedes-Benz   300E      Midsize    61.9

In [82]:
cars_df.loc[cars_df['Price'] == cars_df['Price'].max()][['Manufacturer', 'Model', 'Type', 'Price']] # INSERT CODE HERE

Manufacturer Model     Type  Price
58  Mercedes-Benz  300E  Midsize   61.9

**Q.8** We want to see if our dataset has any missing values. (If it does, we may need to transform them to make the dataset compatible with machine learning algorithms.)

Find all the missing values in the dataset now.

*Sample Output*:

>               Manufacturer    Model     Type       Price
>     58        Mercedes-Benz   300E      Midsize    61.9

In [83]:
cars_df.isna().sum() # INSERT CODE HERE

Manufacturer           4
Model                  1
Type                   3
Min.Price              7
Price                  2
Max.Price              5
MPG.city               9
MPG.highway            2
AirBags                6
DriveTrain             7
Cylinders              5
EngineSize             2
Horsepower             7
RPM                    3
Rev.per.mile           6
Man.trans.avail        5
Fuel.tank.capacity     8
Passengers             2
Length                 4
Wheelbase              1
Width                  6
Turn.circle            5
Rear.seat.room         4
Luggage.room          19
Weight                 7
Origin                 5
Make                   3
dtype: int64

**Q.9** Normally, we'd need to take a closer look at each column containing `null` values before deciding what to do with them. But for the purpose of this notebook, let's just assume we want to get rid of them!

If you ran the above cell correctly, it should show that the `'Model'` column only has one missing value. We shouldn't miss this one too much. **Drop this row now.** Then, check to see how many missing values are in that column.

*Hint: If your code still shows 1 missing value in `'Model'`, try using `inplace`!*

*Output:*

> `# Shape of the dataframe with the missing value 'Model' row gone`

> (92, 27)

In [88]:
# drop the row with a missing value in `Model`
cars_df = cars_df.loc[cars_df['Model'].isna() == False] # INSERT CODE HERE

# check the shape of the dataframe; there should be 92 rows
cars_df.shape

(92, 27)

In [89]:
# check that the missing value from `Model` was removed
cars_df['Model'].isna().sum() # INSERT CODE HERE

0

**Q.10** We can drop entire rows as well. Again, in real life we'd want to investigate before doing this. But this is just for practice! So let's try it out.

If you ran the Q.8 code cell correctly, you should see that the column `'Luggage.room'` has 19 missing values. That's way more than any other column of the dataset!

Get rid of this column now.

*Output:*

> `# Shape of the dataframe with "Luggage.room" gone`

> (92, 26)

In [93]:
# INSERT CODE HERE

cars_df.shape

(92, 26)

42
64
67
75

- create/read df
- describe
- summary
- indexing (loc, iloc)
- sorting
- groupby/agg
- transposing

From MLM:
```python
import numpy
import pandas
myarray = numpy.array([[1, 2, 3], [4, 5, 6]])
rownames = ['a', 'b']
colnames = ['one', 'two', 'three']
mydataframe = pandas.DataFrame(myarray, index=rownames, columns=colnames)
print(mydataframe)
```

EXERCISES: https://www.machinelearningplus.com/python/101-pandas-exercises-python/

## Part 4: Matplotlib

[intro to what matplotlib is, and why we use it]

[what they'll be doing in this section]

- draw simple plot
- create multiple subplots
- piechart
- histogram
- bar plot
- saving a plot

https://jakevdp.github.io/PythonDataScienceHandbook/04.00-introduction-to-matplotlib.html

https://heartbeat.fritz.ai/introduction-to-matplotlib-data-visualization-in-python-d9143287ae39

https://www.tutorialdocs.com/article/python-matplotlib-tutorial.html

EXERCISES: https://www.w3resource.com/graphics/matplotlib/